enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,609 @@
+
+This is the proposed spec of the CRD:
+```yaml
+apiVersion: "persistentip.cni.cncf.io"
+kind: PersistentIP
+metadata:
+  name: my-shiny-vm-persistent-ip
+  namespace: ns1
+  labels:
+    persistentips.cni.cncf.io/owner: VirtualMachine.vm-a.a0790345-4e84-4257-837a-e3d762d191ab
+  ownerReferences:
+  - apiVersion: kubevirt.io/v1
+    kind: VirtualMachine
+    name: vm-a
+    uid: a0790345-4e84-4257-837a-e3d762d191ab
+spec:
+  network: tenantred
+  interface: net2
+status:
+  ip: 10.128.20.8/24
+```
+
+From a CNI perspective, the plugin **must**:
+- **not** delete `PersistentIP`s for encapsulating entities once their pods
+  are deleted. We need to follow the lifecycle of the encapsulating entity,
+  described via the `metadata.ownerReferences` attribute. I.e. if we're
+  deleting a pod where a KubeVirt VM will run, do not remove the `PersistentIP`
+  allocation when processing the CNI delete.
+- retrieve any `PersistentIP`s previously allocated for an encapsulating entity
+  when processing CNI adds for pods implementing said encapsulating entity.
+  I.e. when starting a previously stopped VM, the CNI plugin should check if
+  there are existing `PersistentIP`s for it, and re-use them if found.
+
+Deleting the  `PersistentIP` from the datastore when the pod is deleted will be
+performed by Kubernetes (garbage collection) once the encapsulating object
+(i.e. owner) is removed. In the example above, that `PersistentIP` will be
+garbage collected by Kubernetes when the `VirtualMachine` named `vm-a` with UID
+`a0790345-4e84-4257-837a-e3d762d191ab` is deleted.
+
+### Implementation Details/Notes/Constraints
+
+#### Instructing the IPAM CNI the ADD/DEL requests are on behalf of an owner
+
+We require different behaviors from the IPAM CNI when processing **ADD** events:
+- when an owner is specified, the IPAM CNI plugin **will** try to recover
+  previous allocations for this particular owner. If found, they are re-used.
+  Generated otherwise.
+- when the setup command features an owner, we must persist the IP allocation
+  in the datastore for every IP allocation we generate.
+
+The behavior is also different when processing CNI deletes: the `PersistentIP`
+CRs will not be deleted by the CNI plugin: Kubernetes will garbage collect
+these once the encapsulating object is removed.
+
+The question now becomes how does the client instruct the IPAM CNI plugin this
+particular operation is for a pod owned by another entity ?
+
+The generic way we have derived is via CNI args; the client must specify the
+pod indicating its ownership. Check the following network selection elements
+annotation from a KubeVirt launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "whereabouts-conf",
+    "namespace": "default",
+    "ips": [
+      "192.168.2.13/24"
+    ],
+    "interface": "pod16367aacb67",
+    "cni-args": {
```

Comment:
we'll need to add yet another one, "secondary key" or something similar so we can have a composite key - i.e. stateful set scenario.

Reply:
the "primary key" would be the "ownerName" (or UID) plus the secondary key.